File: frontend/src/types/report.ts
export type ReportStatus =
  | "QUEUED"
  | "ANALYZING"
  | "GENERATING"
  | "COMPLETED"
  | "FAILED";

export type EvidenceLabel =
  | "DIRECTLY_STATED"
  | "DERIVED_FROM_DOCUMENT"
  | "MULTIPLE_SOURCES"
  | "POTENTIAL_GAP"
  | "CONFLICTING_INFORMATION"
  | "NOT_SPECIFIED";

export interface ReportSource {
  index: number;
  document_id?: string | null;
  file_name: string;
  page_number?: number | null;
  section_title?: string | null;
  snippet?: string;
  location_label?: string | null;
  chunk_id?: string | null;
  element_type?: string | null;
}

export interface CitedText {
  text: string;
  source_index?: number | null;
  evidence?: EvidenceLabel;
}

export interface ProfileField {
  label: string;
  value: string;
  source_index?: number | null;
  evidence?: EvidenceLabel;
}

export interface FactRow {
  attribute: string;
  value: string;
  source_index?: number | null;
  evidence?: EvidenceLabel;
}

export interface MetricRow {
  metric: string;
  value: string;
  context?: string;
  source_index?: number | null;
  evidence?: EvidenceLabel;
  calculated?: boolean;
}

export interface MilestoneRow {
  milestone: string;
  date: string;
  source_index?: number | null;
}

export interface RequirementGroup {
  category: string;
  items: CitedText[];
}

export interface EntityGroup {
  category: string;
  items: CitedText[];
}

export interface RiskItem {
  risk: string;
  impact?: string;
  mitigation?: string;
  mitigation_is_recommendation?: boolean;
  category?: string;
  source_index?: number | null;
}

export interface GapItem {
  gap: string;
  impact?: string;
  source_index?: number | null;
}

export interface ConflictItem {
  topic: string;
  statements: CitedText[];
  status?: string;
}

export interface ActionRow {
  action: string;
  owner?: string;
  due?: string;
  source_index?: number | null;
}

export interface DecisionItem {
  decision: string;
  rationale?: string;
  source_index?: number | null;
}

export interface TableSummary {
  title: string;
  summary?: string;
  rows?: string[][];
  source_index?: number | null;
}

export interface FigureRef {
  title: string;
  purpose?: string;
  sheet?: string | null;
  source_index?: number | null;
  description?: string;
}

export interface DocumentProfile {
  name?: ProfileField;
  document_type?: ProfileField;
  revision?: ProfileField;
  author?: ProfileField;
  organization?: ProfileField;
  creation_date?: ProfileField;
  approval_date?: ProfileField;
  effective_date?: ProfileField;
  page_count?: ProfileField;
  language?: ProfileField;
  status?: ProfileField;
}

export interface ReportContent {
  title?: string;
  document_names?: string[];
  document_count?: number;
  limitations?: string | null;
  executive_summary?: string;
  purpose?: string;
  top_findings?: CitedText[];
  document_profile?: DocumentProfile;
  key_facts?: FactRow[];
  requirements?: RequirementGroup[];
  metrics?: MetricRow[];
  milestones?: MilestoneRow[];
  entities?: EntityGroup[];
  tables?: TableSummary[];
  figures?: FigureRef[];
  risks?: RiskItem[];
  gaps?: GapItem[];
  conflicts?: ConflictItem[];
  actions?: ActionRow[];
  decisions?: DecisionItem[];
  document_actions?: CitedText[];
  observations?: string[];
  cross_document_findings?: ConflictItem[];
  sources?: ReportSource[];
}

export interface ReportPayload {
  reportId: string;
  status: ReportStatus;
  reportType?: string;
  title: string;
  documentCount: number;
  documentNames: string[];
  pageCount?: number | null;
  generatedAt?: string | null;
  pdfUrl?: string | null;
  downloadName?: string | null;
  limitations?: string | null;
  errorMessage?: string | null;
  content?: ReportContent | null;
  sections?: ReportContent | null;
}

export interface ReportMedia {
  media_type: "document_report";
  report_id: string;
  title?: string;
  page_count?: number | null;
  document_names?: string[];
  download_name?: string | null;
  limitations?: string | null;
}

export function isReportMedia(item: unknown): item is ReportMedia {
  return (
    typeof item === "object" &&
    item !== null &&
    (item as { media_type?: string }).media_type === "document_report" &&
    typeof (item as { report_id?: string }).report_id === "string"
  );
}

File: backend/app/schemas/report.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field


class ReportType(str, Enum):
    FULL_DOCUMENT_REPORT = "FULL_DOCUMENT_REPORT"
    # Future: SUMMARY, TECHNICAL, RISKS, COMPLIANCE, SCHEDULE, BOQ


class ReportStatus(str, Enum):
    QUEUED = "QUEUED"
    ANALYZING = "ANALYZING"
    GENERATING = "GENERATING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class EvidenceLabel(str, Enum):
    DIRECTLY_STATED = "DIRECTLY_STATED"
    DERIVED_FROM_DOCUMENT = "DERIVED_FROM_DOCUMENT"
    MULTIPLE_SOURCES = "MULTIPLE_SOURCES"
    POTENTIAL_GAP = "POTENTIAL_GAP"
    CONFLICTING_INFORMATION = "CONFLICTING_INFORMATION"
    NOT_SPECIFIED = "NOT_SPECIFIED"


class Priority(str, Enum):
    CRITICAL = "CRITICAL"
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"


NOT_SPECIFIED = "Not specified in the document."


class SourceRef(BaseModel):
    index: int
    document_id: Optional[str] = None
    file_name: str
    page_number: Optional[int] = None
    section_title: Optional[str] = None
    snippet: str = ""
    location_label: Optional[str] = None
    chunk_id: Optional[str] = None
    element_type: Optional[str] = None


class CitedText(BaseModel):
    text: str
    source_index: Optional[int] = None
    evidence: EvidenceLabel = EvidenceLabel.DIRECTLY_STATED
    priority: Priority = Priority.HIGH


class ProfileField(BaseModel):
    label: str
    value: str = NOT_SPECIFIED
    source_index: Optional[int] = None
    evidence: EvidenceLabel = EvidenceLabel.NOT_SPECIFIED


class FactRow(BaseModel):
    attribute: str
    value: str
    source_index: Optional[int] = None
    evidence: EvidenceLabel = EvidenceLabel.DIRECTLY_STATED
    priority: Priority = Priority.HIGH


class MetricRow(BaseModel):
    metric: str
    value: str
    context: str = ""
    source_index: Optional[int] = None
    evidence: EvidenceLabel = EvidenceLabel.DIRECTLY_STATED
    calculated: bool = False
    priority: Priority = Priority.HIGH


class RequirementGroup(BaseModel):
    category: str
    items: list[CitedText] = Field(default_factory=list)


class MilestoneRow(BaseModel):
    milestone: str
    date: str
    source_index: Optional[int] = None
    evidence: EvidenceLabel = EvidenceLabel.DIRECTLY_STATED
    priority: Priority = Priority.HIGH


class EntityGroup(BaseModel):
    category: str
    items: list[CitedText] = Field(default_factory=list)


class TableSummary(BaseModel):
    title: str
    summary: str
    rows: list[list[str]] = Field(default_factory=list)
    source_index: Optional[int] = None
    priority: Priority = Priority.MEDIUM


class FigureRef(BaseModel):
    title: str
    purpose: str = NOT_SPECIFIED
    sheet: Optional[str] = None
    source_index: Optional[int] = None
    description: str = ""
    priority: Priority = Priority.MEDIUM


class RiskItem(BaseModel):
    risk: str
    impact: str = NOT_SPECIFIED
    mitigation: str = NOT_SPECIFIED
    mitigation_is_recommendation: bool = False
    category: str = ""
    source_index: Optional[int] = None
    evidence: EvidenceLabel = EvidenceLabel.DIRECTLY_STATED
    priority: Priority = Priority.CRITICAL


class GapItem(BaseModel):
    gap: str
    impact: str = ""
    source_index: Optional[int] = None
    evidence: EvidenceLabel = EvidenceLabel.POTENTIAL_GAP
    priority: Priority = Priority.HIGH


class ConflictItem(BaseModel):
    topic: str
    statements: list[CitedText] = Field(default_factory=list)
    status: str = "Requires clarification."
    evidence: EvidenceLabel = EvidenceLabel.CONFLICTING_INFORMATION
    priority: Priority = Priority.CRITICAL


class ActionRow(BaseModel):
    action: str
    owner: str = "Not specified"
    due: str = "Not specified"
    source_index: Optional[int] = None
    evidence: EvidenceLabel = EvidenceLabel.DIRECTLY_STATED
    priority: Priority = Priority.HIGH


class DecisionItem(BaseModel):
    decision: str
    rationale: str = NOT_SPECIFIED
    source_index: Optional[int] = None
    evidence: EvidenceLabel = EvidenceLabel.DIRECTLY_STATED
    priority: Priority = Priority.HIGH


class DocumentProfile(BaseModel):
    name: ProfileField = Field(default_factory=lambda: ProfileField(label="Document name"))
    document_type: ProfileField = Field(
        default_factory=lambda: ProfileField(label="Document type")
    )
    revision: ProfileField = Field(default_factory=lambda: ProfileField(label="Revision"))
    author: ProfileField = Field(default_factory=lambda: ProfileField(label="Author"))
    organization: ProfileField = Field(
        default_factory=lambda: ProfileField(label="Organization")
    )
    creation_date: ProfileField = Field(
        default_factory=lambda: ProfileField(label="Creation date")
    )
    approval_date: ProfileField = Field(
        default_factory=lambda: ProfileField(label="Approval date")
    )
    effective_date: ProfileField = Field(
        default_factory=lambda: ProfileField(label="Effective date")
    )
    page_count: ProfileField = Field(default_factory=lambda: ProfileField(label="Page count"))
    language: ProfileField = Field(default_factory=lambda: ProfileField(label="Language"))
    status: ProfileField = Field(default_factory=lambda: ProfileField(label="Document status"))


class ReportContent(BaseModel):
    title: str = "BuildLens Document Report"
    report_type: str = ReportType.FULL_DOCUMENT_REPORT.value
    document_names: list[str] = Field(default_factory=list)
    document_count: int = 0
    limitations: Optional[str] = None
    executive_summary: str = ""
    purpose: str = ""
    top_findings: list[CitedText] = Field(default_factory=list)
    document_profile: DocumentProfile = Field(default_factory=DocumentProfile)
    key_facts: list[FactRow] = Field(default_factory=list)
    requirements: list[RequirementGroup] = Field(default_factory=list)
    metrics: list[MetricRow] = Field(default_factory=list)
    milestones: list[MilestoneRow] = Field(default_factory=list)
    entities: list[EntityGroup] = Field(default_factory=list)
    tables: list[TableSummary] = Field(default_factory=list)
    figures: list[FigureRef] = Field(default_factory=list)
    risks: list[RiskItem] = Field(default_factory=list)
    gaps: list[GapItem] = Field(default_factory=list)
    conflicts: list[ConflictItem] = Field(default_factory=list)
    actions: list[ActionRow] = Field(default_factory=list)
    decisions: list[DecisionItem] = Field(default_factory=list)
    document_actions: list[CitedText] = Field(default_factory=list)
    observations: list[str] = Field(default_factory=list)
    cross_document_findings: list[ConflictItem] = Field(default_factory=list)
    sources: list[SourceRef] = Field(default_factory=list)


class CreateReportRequest(BaseModel):
    documentIds: list[str] = Field(default_factory=list)
    type: ReportType = ReportType.FULL_DOCUMENT_REPORT
    sessionId: Optional[str] = None
    projectId: Optional[str] = None


class ReportResponse(BaseModel):
    reportId: str
    status: str
    reportType: str = ReportType.FULL_DOCUMENT_REPORT.value
    title: str = "Document Report"
    documentCount: int = 0
    documentNames: list[str] = Field(default_factory=list)
    pageCount: Optional[int] = None
    generatedAt: Optional[datetime] = None
    pdfUrl: Optional[str] = None
    downloadName: Optional[str] = None
    limitations: Optional[str] = None
    errorMessage: Optional[str] = None
    sections: Optional[dict[str, Any]] = None
    content: Optional[dict[str, Any]] = None
    progress: Optional[str] = None
